libs/api/libbpf_program.cpp

```diff
@@ -422,3 +420,130 @@ bpf_prog_bind_map(int prog_fd, int map_fd, const struct bpf_prog_bind_opts* opts
 
     return libbpf_result_err(ebpf_program_bind_map(prog_fd, map_fd));
 }
+
+static int
+__bpf_set_link_xdp_fd_replace(int ifindex, int fd, int old_fd, __u32 flags)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+
+    // On Linux, the logic to detach the older program and attach the
+    // new program is present in kernel (the flag XDP_FLAGS_REPLACE is
+    // passed to the kernel).  Thus, we should consider moving this
+    // logic to the execution context.
+    if (flags & XDP_FLAGS_REPLACE) {
+        // Look up the old program info to get the program ID.
+        ebpf_id_t program_id = 0;
+        if (old_fd == 0 || old_fd == ebpf_fd_invalid) {
+            int err = bpf_xdp_query_id(ifindex, 0, &program_id);
+            if ((err < 0) && (errno != ENOENT)) {
+                return err;
+            }
+            // We found the right program_id.
+        } else {
+            struct bpf_prog_info prog_info;
+            uint32_t info_len = sizeof(prog_info);
+            int err = bpf_obj_get_info_by_fd(old_fd, &prog_info, &info_len);
+            if (err < 0) {
+                return err;
+            }
+
+            // Verify that the program is actually an XDP program.
+            if (prog_info.type != BPF_PROG_TYPE_XDP) {
+                return libbpf_err(-EINVAL);
+            }
```

Comment:
This is kind of interesting that this program may not be associated with the input ifindex - granted that would be an error on the calling apps part - but possible. Just like we are verifying the prog type is XDP, should be also check if the ifindex matches? But for that link_info is needed.

Why not call `bpf_xdp_query_id` irrespective of input old_fd. If the old_fd is specified, we can perform an additional check if the returned prog_info matches the prog_info corresponding to the old_fd ?

Reply:
My thought was that if we were to allow multiple programs to be attached to the same ifindex, then bpf_xdp_query_id would only return one of them, but this API would still allow replacing any one of them, not just the one returned by bpf_xdp_query_id, by specifying which one to replace.